_posts/2021-05-14-Spring-Boot-TestConfiguration.md

```diff
@@ -0,0 +1,215 @@
+import org.springframework.web.reactive.function.client.WebClient;
+
+@Configuration
+public class WebClientConfiguration {
+        @Bean
+        public WebClient getWebClient(final WebClient.Builder builder,
+               @Value("${data.service.endpoint:https://google.com}") 
+                  final String url) {
+            WebClient webClient = builder.baseUrl(url)
+            .defaultHeader(HttpHeaders.CONTENT_TYPE, 
+             MediaType.APPLICATION_JSON_VALUE)
+             // more configurations and customizations
+             .build();
+                System.out.println("WebClient Instance Created During Testing: " 
+                + webClient.toString());
+                return webClient;
+        }
+}
+```
+
+### Running the application
+Let us run the application using `mvn spring-boot:run` to verify that the WebClient created by the `@Configuration` class is injected into the service.
+
+In the console log, we can observe that `webClient.toString()` in both configuration and service class prints the same value:
+
+```shell
+WebClient Instance Created During Testing: org.springframework.web.reactive.function.client.DefaultWebClient@e4533f3a
+WebClient instance org.springframework.web.reactive.function.client.DefaultWebClient@e4533f3a
+```
+
+### Implementing Test using @Import
+Let us implement a simple unit test that imports the @TestConfiguration to overwrite the WebClient and inject the same to the service class:
+
+```java
+@SpringBootTest
+@Import(WebClientTestConfiguration.class)
+@TestPropertySource(locations="classpath:test.properties")
+class TestConfigurationExampleAppTests {
+ @Autowired
+ private DataService dataService;
+ @Test
+ void contextLoads() {
+ }
+}
+```
+
+### Implementing Test using Static Inner Class
+Let us implement a simple unit test that imports the `@TestConfiguration` declared as a static inner class. In this case, we don't need to instruct the Spring Context to load the `@TestConfiguration` class:
+
+```java
+@SpringBootTest
+@TestPropertySource(locations="classpath:test.properties")
+public class UsingStaticInnerTestConfiguration {
+ @Autowired
+ private DataService dataService;
+
+ @Test
+ void contextLoads() {
+ }
+
+ @TestConfiguration
+ public static class WebClientConfiguration {
+ @Bean
+ public WebClient getWebClient(final WebClient.Builder builder) {
+ WebClient webClient = builder
+ .baseUrl("http://localhost")
+ .build();
+ System.out.println("WebClient Instance Created During Testing, " +
+ "using static inner class: " + webClient.toString());
+ return webClient;
+ }
+ }
+}
+```
+
+When we run the above JUnit test, we can observe that the `@Configuration` class is not called to create the WebClient, but the @TestConfiguration class is called to create the WebClient instance.
+
+We can compare and observe the webClient.toString() method in both the `@TestConfiguration` class and the service implementation, which are found to be the same.
```

Comment:
Rephrase this. Not very clear.

Reply:
This section has been rewritten. Please check once again.